Fix several bugs around LIKE/ILIKE and regexes

ggevay opened the pull request:
(Doc changes are only in the first commit. They are described [here](https://materializeinc.slack.com/archives/CPNFV9CVB/p1688728226054109).)

This is a fix for most of the bugs described here: https://github.com/MaterializeInc/materialize/issues/18494#issuecomment-1624061454
The only thing I didn't fix is the serde serialization, but this is not currently used for anything important (see comments in the code).

The approach for fixing the protobuf serialization is to add the pattern string and the `case_insensitive` flag in our `Regex` wrapper struct, and write a serializer for our wrapper struct that serializes these fields (but not the actual compiled regex), and rebuild the compiled regex from these fields when deserializing.

The last commit adds a lot more tests for all this functionality. This involves testing with 3 levels of const folding:
- Full const folding, where all mentions of regexes or LIKE disappear.
- Partial const folding, where the binary (or variadic) version of regex functions or LIKE functions are changed to unary versions that hardcode the regex (so that it doesn't need to be compiled at every invocation).
- No const folding, where the pattern comes from the data.

FYI @mjibson: we are probably gonna have some merge conflicts with https://github.com/MaterializeInc/materialize/pull/20585.

### Motivation

  * This PR fixes a recognized bug and several other bugs, see [here](https://github.com/MaterializeInc/materialize/issues/18494#issuecomment-1624061454).

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

ggevay commented:
(Fixing the lint issue now.)

nrainer-materialize commented:
I assume this will also fix the output inconsistency. 😀 Could you please enable some more tests by deleting the following code:

`./misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py`, lines `189`-`195`:
```
        if db_function.function_name_in_lower_case in {"regexp_match"}:
            if len(expression.args) == 3 and isinstance(
                expression.args[2], EnumConstant
            ):
                # This is a regexp_match function call with case-insensitive configuration.
                # https://github.com/MaterializeInc/materialize/issues/18494
                return YesIgnore("#18494")
```

`inconsistency_ignore_filter.py`, lines `213`-`216`:
```
        # https://github.com/MaterializeInc/materialize/issues/18494
        if db_operation.pattern in {"$ ~* $", "$ !~* $"}:
            return YesIgnore("#18494")
```

Thank you!

ggevay commented:
@philip-stoev, could you please help with fixing the failing regex test in `Checks + DROP/CREATE replica`? I'm very confused:
- The test is running into a timeout at the `UPDATE update_table SET f2 = f2 * 1000;`. However, I think the problem might be something in the last test in `regex.py`, which is shortly before that (since the PR is about regexes), but I really don't understand how this test framework works.
- I looked at the last test in `regex.py`, but I don't see a reason why the output of those commands should change from this PR. If I run those commands locally, then the output is the same as the test assumes.
- There is one thing suspicious in that test: Is it possible that it is relying on some unstable row ordering? (But I don't see a change in the row order when I run those commands locally.)

nrainer-materialize commented on doc/user/content/sql/functions/_index.md:
```diff
@@ -102,6 +102,10 @@ Operator | Computes
 Operator | Computes
 ---------|---------
 <code>&vert;&vert;</code> | Concatenation
+`~~` | Matches LIKE pattern case sensitively, using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`~~*` | Matches LIKE pattern case insensitively (ILIKE), using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`!~~` | Does not match LIKE pattern case sensitively, using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`!~~*` | Does not match LIKE pattern case insensitively (ILIKE), using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
```
Are these just synonyms for `LIKE` / `ILIKE`?

mjibson replied:
Yes: https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-LIKE

mjibson commented on doc/user/content/sql/functions/_index.md:
```diff
@@ -102,6 +102,10 @@ Operator | Computes
 <code>&vert;&vert;</code> | Concatenation
+`~~` | Matches LIKE pattern case sensitively, using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`~~*` | Matches LIKE pattern case insensitively (ILIKE), using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`!~~` | Does not match LIKE pattern case sensitively, using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
```
```suggestion
`!~~` | Matches NOT LIKE pattern case sensitively, using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
```

ggevay replied:
Yeah, these were also bothering me a bit... I've changed all of them now.

mjibson commented on doc/user/content/sql/functions/_index.md:
```diff
@@ -102,6 +102,10 @@ Operator | Computes
+`~~` | Matches LIKE pattern case sensitively, using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`~~*` | Matches LIKE pattern case insensitively (ILIKE), using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`!~~` | Does not match LIKE pattern case sensitively, using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
+`!~~*` | Does not match LIKE pattern case insensitively (ILIKE), using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
```
```suggestion
`!~~*` | Matches NOT LIKE pattern case insensitively (ILIKE), using [SQL LIKE matching](https://www.postgresql.org/docs/13/functions-matching.html#FUNCTIONS-LIKE)
```

ggevay replied:
changed

mjibson commented on src/expr/src/relation/func.rs:
```diff
@@ -1898,7 +1898,7 @@ impl RustType<ProtoAnalyzedRegex> for AnalyzedRegex {
 
 impl AnalyzedRegex {
     pub fn new(s: &str) -> Result<Self, regex::Error> {
```
```suggestion
    pub fn new(s: String) -> Result<Self, regex::Error> {
```

Using a String might avoid an allocation if the caller already has a String.

ggevay replied:
Thx, changed

mjibson commented on src/expr/src/scalar/like_pattern.rs:
```diff
@@ -129,17 +128,13 @@ impl RustType<ProtoMatcher> for Matcher {
             pattern: self.pattern.clone(),
             case_insensitive: self.case_insensitive,
-            matcher_impl: Some(self.matcher_impl.into_proto()),
+            //matcher_impl: Some(self.matcher_impl.into_proto()),
```
```suggestion
```

ggevay replied:
Sorry, fixed

mjibson commented on src/repr/src/adt/regex.proto:
```diff
@@ -12,5 +12,6 @@ syntax = "proto3";
 
 message ProtoRegex {
-    string regex = 1;
+    string pattern = 1;
```
Is name changing in a proto safe?

ggevay replied:
Well, the linked thread says that "renaming fields is a wire compatible change". (The field name in the `ProtoRegex` struct changes, but I've of course changed the Rust code that is dealing with `ProtoRegex` directly.)

Also, we don't currently need backwards-compatibility for protobuf serialization, as far as I know.

aalexandrov replied:
Sorry, I meant to say "I think so".

mjibson commented on src/sql/src/func.rs:
```diff
@@ -3847,7 +3847,7 @@ pub static OP_IMPLS: Lazy<BTreeMap<&'static str, Func>> = Lazy::new(|| {
             params!(Char, String) => Operation::binary(|ecx, lhs, rhs| {
                 let length = ecx.scalar_type(&lhs).unwrap_char_length();
                 Ok(lhs.call_unary(UnaryFunc::PadChar(func::PadChar { length }))
-                    .call_binary(rhs, IsLikeMatch { case_insensitive: false })
```
Wow! Uh, so this PR will do the opposite behavior for all installed views? We might need to look through customer views and inform any of these.

ggevay replied:
Yes, ILIKE was almost completely broken due to all these bugs. I'll try to write a query that we can run against all customer environments to see if anyone was affected without them knowing.

ggevay commented:
Tests are failing because it seems [we are totally using the Deserialize/Serialize impls of `Regex` for important stuff](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1689620877396529?thread_ts=1688665161.517769&cid=C02PPB50ZHS), so then we can't postpone fixing these... I'll fix them tomorrow. Converting the PR back to draft for now.

ggevay commented on src/repr/src/adt/regex.rs:
```diff
@@ -39,21 +41,47 @@ include!(concat!(env!("OUT_DIR"), "/mz_repr.adt.regex.rs"));
+    // Anyhow, this is not so urgent to fix, because it is only used by non-essential things AFAIK:
+    // - `MzReflect`
+    // - `EXPLAIN AS JSON` (through `DisplayJson`)
+    // For now, we just skip serializing it, to avoid putting incorrect information in
```
Yes, working on this now.

ggevay replied:
Fixed

ggevay commented on src/repr/src/adt/regex.rs:
```diff
@@ -39,21 +41,47 @@ include!(concat!(env!("OUT_DIR"), "/mz_repr.adt.regex.rs"));
+    // our wrapping [Regex] struct, which _will_ show the `case_insensitive` field. Also note that
+    // EXPLAIN AS JSON does only serialization, but not deserialization.
+    // (see also the commented out test `regex_serde_case_insensitive`)
+    #[serde(with = "serde_regex", skip_serializing)]
```
Yes, I'll throw out `serde_regex`, and write a manual Serialize/Deserialize for our wrapper struct. The manual code will reconstruct the wrapped `Regex` from the other fields.

ggevay replied:
Done

ggevay commented:
I addressed all comments, and all tests are passing. 

As Matt suggested, I’ll try to see if any customer environments might be affected by the buggy ILIKE, so that we can notify them.

ggevay commented:
Ok, the only affected user is Centerfield, and [Seth says he is going to send them a msg before the release goes out](https://materializeinc.slack.com/archives/C04LQV4KUR4/p1689865411198419?thread_ts=1689865248.166389&cid=C04LQV4KUR4).
